sdk-tools/src/sign.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright (c) 2023 The MobileCoin Foundation
+
+//! Builder wrapper around SgxSign.
+
+use std::{
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+/// This structure wraps the execution of sgx_sign commands.
+#[derive(Clone, Debug)]
+pub struct SgxSign {
+    /// The path to the sgx_sign executable.
+    sgx_sign_path: PathBuf,
+    /// Whether to ignore the presence of relocations in the enclave shared
+    /// object.
+    ignore_relocation_error: bool,
+    /// Whether to ignore .init sections in the enclave.
+    ignore_init_section_error: bool,
+    /// Whether to re-sign a previously signed enclave (default: false)
+    resign: bool,
+}
+
+impl Default for SgxSign {
+    /// Create a new SGX signing utility from the current environment.
+    fn default() -> Self {
+        let bin_dir = mc_sgx_core_build::sgx_bin_x64_dir();
+
+        SgxSign::from(bin_dir.join("sgx_sign"))
+    }
+}
+
+impl SgxSign {
+    /// Relocations are generally forbidden in the enclave shared object, this
+    /// tells the `sgx_sign` utility to ignore those errors.
+    #[must_use]
+    pub fn allow_relocations(mut self, allow: bool) -> Self {
+        self.ignore_relocation_error = allow;
+        self
+    }
+
+    /// Whether or not to allow .init sections in the enclave.
+    #[must_use]
+    pub fn allow_init_sections(mut self, allow: bool) -> Self {
+        self.ignore_init_section_error = allow;
+        self
+    }
+
+    /// Whether to re-sign a previously signed enclave (default: false)
+    #[must_use]
+    pub fn allow_resign(mut self, allow: bool) -> Self {
+        self.resign = allow;
+        self
+    }
+
+    /// Generate the command to sign the given enclave object with the given
+    /// private key and write the resulting enclave to the given path. Note
+    /// that online signatures are inherently insecure.
+    pub fn sign(
+        &mut self,
+        unsigned_enclave: &Path,
+        config_path: &Path,
+        private_key: &Path,
+        output_enclave: &Path,
```

Comment:
I think the problems you're mentioning with changing this to `impl AsRef<Path>` are actually significant.

Firstly, with just `impl AsRef<Path>` that means "take ownership of a type which can be referenced as a path." This is why you need to change callers to use `as_ref()`, which is super awkward. This will also entail a `memmove()` if you just give it a `PathBuf` If you wanted to do this, it should be `&impl AsRef<Path>`, because we don't actually want to take ownership here.

The second issue is, as noted, the single `P` forces everything to be the same type, so you'd need `P1`, `P2`, `P3` to be useful (this is what `impl AsRef<Path>` desugars to, AFAIK).

Reply:
Looking at https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3ecd960e21aa6cdd8dc328cc216ad275 
It seems that `&impl AsRef<Path>` is more restrictive, it forces all callers to pass in a reference, which prevents
```rust
call_function(Path::new("stuff");
```
And the `impl AsRef<Path>` allows references as well as owning.  The `as_ref()` should only need to be done in the callee.

Reply:
Hmm, OK, complaints withdrawn :-)